internal/featuretests/v3/tlsroute_test.go

```diff
@@ -0,0 +1,170 @@
+	defer done()
+
+	//s1 := &v1.Secret{
+	//	ObjectMeta: metav1.ObjectMeta{
+	//		Name:      "secret",
+	//		Namespace: "default",
+	//	},
+	//	Type: "kubernetes.io/tls",
+	//	Data: featuretests.Secretdata(featuretests.CERTIFICATE, featuretests.RSA_PRIVATE_KEY),
+	//}
+
+	svc := fixture.NewService("correct-backend").
+		WithPorts(v1.ServicePort{Port: 80, TargetPort: intstr.FromInt(8080)})
+
+	//rh.OnAdd(s1)
+	rh.OnAdd(svc)
+
+	rh.OnAdd(&gatewayapi_v1alpha1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "contour",
+			Namespace: "projectcontour",
+		},
+		Spec: gatewayapi_v1alpha1.GatewaySpec{
+			Listeners: []gatewayapi_v1alpha1.Listener{{
+				Port:     443,
+				Protocol: "TLS",
+				Routes: gatewayapi_v1alpha1.RouteBindingSelector{
+					Namespaces: &gatewayapi_v1alpha1.RouteNamespaces{
+						From: routeSelectTypePtr(gatewayapi_v1alpha1.RouteSelectAll),
+					},
+					Kind: dag.KindTLSRoute,
+				},
+			}},
+		},
+	})
+
+	route1 := &gatewayapi_v1alpha1.TLSRoute{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "basic",
+			Namespace: "default",
+		},
+		Spec: gatewayapi_v1alpha1.TLSRouteSpec{
+			Rules: []gatewayapi_v1alpha1.TLSRouteRule{{
+				Matches: []gatewayapi_v1alpha1.TLSRouteMatch{{
+					SNIs: []gatewayapi_v1alpha1.Hostname{
+						"tcp.projectcontour.io",
+					},
+				}},
+				ForwardTo: []gatewayapi_v1alpha1.RouteForwardTo{{
+					ServiceName: pointer.StringPtr("correct-backend"),
+					Port:        gatewayPort(80),
+				}},
+			}},
+		},
+	}
+
+	rh.OnAdd(route1)
+
+	c.Request(listenerType).Equals(&envoy_discovery_v3.DiscoveryResponse{
+		Resources: resources(t,
+			&envoy_listener_v3.Listener{
+				Name:    "ingress_https",
+				Address: envoy_v3.SocketAddress("0.0.0.0", 8443),
+				FilterChains: []*envoy_listener_v3.FilterChain{{
+					Filters: envoy_v3.Filters(
+						tcpproxy("ingress_https", "default/correct-backend/80/da39a3ee5e"),
+					),
+					FilterChainMatch: &envoy_listener_v3.FilterChainMatch{
+						ServerNames: []string{"tcp.projectcontour.io"},
+					},
+				}},
+				ListenerFilters: envoy_v3.ListenerFilters(
+					envoy_v3.TLSInspector(),
+				),
+				SocketOptions: envoy_v3.TCPKeepaliveSocketOptions(),
+			},
+			staticListener(),
+		),
+		TypeUrl: listenerType,
+	})
+
+	// check that ingress_http is empty
+	c.Request(routeType).Equals(&envoy_discovery_v3.DiscoveryResponse{
+		Resources: resources(t,
+			envoy_v3.RouteConfiguration("ingress_http"),
+		),
+		TypeUrl: routeType,
+	})
+
+	// Route2 doesn't define any SNIs, so this should become the default backend.
+	route2 := &gatewayapi_v1alpha1.TLSRoute{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "basic",
+			Namespace: "default",
+		},
+		Spec: gatewayapi_v1alpha1.TLSRouteSpec{
+			Rules: []gatewayapi_v1alpha1.TLSRouteRule{{
+				ForwardTo: []gatewayapi_v1alpha1.RouteForwardTo{{
+					ServiceName: pointer.StringPtr("correct-backend"),
+					Port:        gatewayPort(80),
+				}},
+			}},
+		},
+	}
+
+	rh.OnUpdate(route1, route2)
```

Comment:
maybe add a test for one rule with no SNI and another with an SNI?

Reply:
👍